pyrevolve/schedulers/__init__.py

```diff
@@ -0,0 +1,544 @@
+                self.old_capo = h_op.index[1]
+
+            self.index = h_op.index
+
+    def storageIndex(self):
+        if (
+            (self.type == Action.TAKESHOT)
+            or (self.type == Action.RESTORE)
+            or (self.type == Action.CPDEL)
+        ):
+            return self.index[0]
+        else:
+            return 0
+
+    def __repr__(self):
+        return json.dumps(
+            dict(
+                {
+                    "type": self.type_names[self.type],
+                    "capo": self.capo,
+                    "old_capo": self.old_capo,
+                    "ckp": self.ckp,
+                    "st_idx": self.storageIndex(),
+                }
+            )
+        )
+
+
+class Scheduler(object):
+    """
+    Abstract base class for scheduler implementations.
+    This class defines the scheduler interface used by
+    pyrevolve FSM.
+    """
+
+    __metaclass__ = ABCMeta
+
+    def __init__(self, n_checkpoints, n_timesteps):
+        super().__init__()
+        self.n_timesteps = n_timesteps
+        self.n_checkpoints = n_checkpoints
+
+    @abstractmethod
+    def next(self):
+        return NotImplemented
+
+    @property
+    def capo(self):
+        return 0
+
+    @property
+    def old_capo(self):
+        return 0
+
+    @property
+    def cp_pointer(self):
+        return 0
+
+
+class CRevolve(Scheduler):
+    """
+    Scheduler class based on the CPP implementation of
+    the traditional Revolve Algorithm
+    """
+
+    translations = {
+        cr.Action.advance: Action.ADVANCE,
+        cr.Action.takeshot: Action.TAKESHOT,
+        cr.Action.restore: Action.RESTORE,
+        cr.Action.firstrun: Action.LASTFW,
+        cr.Action.youturn: Action.REVERSE,
+        cr.Action.terminate: Action.TERMINATE,
+    }
+
+    def __init__(self, n_checkpoints, n_timesteps):
+        super().__init__(n_checkpoints, n_timesteps)
+        self.revolve = cr.CRevolve(n_checkpoints, n_timesteps, None)
+        self.__revstart_action = None
+
+    def next(self):
+        if self.__revstart_action is None:
+            ca = CAction(
+                action_type=self.translations[self.revolve.revolve()],
+                capo=self.capo,
+                old_capo=self.old_capo,
+                ckp=self.cp_pointer,
+            )
+            if ca.type is Action.LASTFW:
+                self.__revstart_action = CAction(
+                    action_type=Action.REVSTART,
+                    capo=self.capo,
+                    old_capo=self.old_capo,
+                    ckp=self.cp_pointer,
+                )
+        else:
+            ca = self.__revstart_action
+            self.__revstart_action = None
+        return ca
+
+    @property
+    def capo(self):
+        return self.revolve.capo
+
+    @property
+    def old_capo(self):
+        return self.revolve.oldcapo
+
+    @property
+    def cp_pointer(self):
+        return self.revolve.check
+
+
+class HRevolve(Scheduler):
```

Comment:
Can we please move class `HRevolve` to `schedulers/hrevolve.py` and similarly for `CRevolve`, importing them into `__init__.py` wherever appropriate?

Reply:
HRevolve and CRevolve depend on Action and Scheduler base-classes which are defined into the _ _init_ _.py file. If I move them into hrevolve.py and crevolve.py files I would have to import _ _init_ _.py not the other way round. Then I would have to import them into pyrevolve.py. I can create a third file called (I don't know) schedbase.py and put Scheduler (base-classe), Architecture, and Action classes, import this module into hrevolve.py/crevolve.py and then import them all into _ _init_ _.py. It would be a mess. What do you think?

Reply:
I like the idea of having a `base.py` that contains these base types in a separate file. I don't understand why you think that is a mess? I feel like that would leave things more organised?

Reply:
I'll try to organize it then.